File: packages/relay-core/src/handlers/certs.ts
/**
 * Certificate handlers for JWKS distribution.
 */

import { Hono } from "hono";
import type { RelayConfig, TenantConfig } from "../config/types.js";

/**
 * JWK structure for certificates.
 */
interface JWK {
  kty: string;
  crv: string;
  kid: string;
  x?: string;
  d?: string;
}

/**
 * JWKS structure.
 */
interface JWKS {
  keys: JWK[];
}

/**
 * Derive Ed25519 public key from seed.
 * Note: This requires Web Crypto API with Ed25519 support or a polyfill.
 */
async function deriveEd25519PublicKey(_seed: Uint8Array): Promise<Uint8Array> {
  // Web Crypto API doesn't support Ed25519 key derivation in all environments
  // For now, we throw an error and expect the JWKS to already have public keys
  throw new Error(
    "Ed25519 public key derivation not supported. Ensure JWKS contains public keys."
  );
}

/**
 * Base64URL decode.
 */
function base64UrlDecode(str: string): Uint8Array {
  // Add padding if needed
  const padded = str + "=".repeat((4 - (str.length % 4)) % 4);
  const base64 = padded.replace(/-/g, "+").replace(/_/g, "/");
  const binary = atob(base64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

/**
 * Base64URL encode.
 */
function base64UrlEncode(bytes: Uint8Array): string {
  let binary = "";
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

/**
 * Compute SHA256 hash and return first 16 bytes as hex.
 */
async function computeETag(data: Uint8Array): Promise<string> {
  const hashBuffer = await crypto.subtle.digest("SHA-256", data);
  const hashArray = Array.from(new Uint8Array(hashBuffer));
  return (
    '"' +
    hashArray
      .slice(0, 16)
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("") +
    '"'
  );
}

/**
 * Redact private keys from JWKS.
 */
async function redactJWKSPrivateKeys(jwksJson: string): Promise<string> {
  const jwks: JWKS = JSON.parse(jwksJson);

  if (!jwks.keys || jwks.keys.length === 0) {
    throw new Error("invalid jwks: keys is missing or empty");
  }

  for (const key of jwks.keys) {
    if (key.d) {
      // If we have a private key but no public key, we need to derive it
      // For now, just remove the private key and hope X is already present
      if (!key.x) {
        // Try to derive public key from seed (not supported in all environments)
        try {
          const seed = base64UrlDecode(key.d);
          const pubKey = await deriveEd25519PublicKey(seed);
          key.x = base64UrlEncode(pubKey);
        } catch {
          throw new Error(
            `Cannot derive public key for ${key.kid}. Ensure JWKS contains public keys.`
          );
        }
      }
      delete key.d;
    }
  }

  return JSON.stringify(jwks);
}

/**
 * Find tenant by allowed domain.
 */
function findTenant(
  tenants: TenantConfig[] | undefined,
  domain: string
): TenantConfig | undefined {
  return tenants?.find(
    (t) => t.allowedDomain.toLowerCase() === domain.toLowerCase()
  );
}

/**
 * Create certificate handlers.
 */
export function createCertsHandlers(config: RelayConfig): Hono {
  const app = new Hono();

  /**
   * GET /relay/certs/:domain - Get public JWKS for a tenant.
   */
  app.get("/relay/certs/:domain", async (c) => {
    const domain = c.req.param("domain")?.trim();
    if (!domain) {
      return c.text("domain is required", 400);
    }

    const tenant = findTenant(config.tenants, domain);
    if (!tenant) {
      return c.text("tenant not found", 404);
    }

    // Get JWKS from tenant config (extended tenant config would have jwks field)
    const jwks = (tenant as TenantConfig & { jwks?: string }).jwks;
    if (!jwks) {
      return c.text("tenant jwks is empty", 500);
    }

    try {
      const redacted = await redactJWKSPrivateKeys(jwks);
      const encoder = new TextEncoder();
      const data = encoder.encode(redacted);
      const etag = await computeETag(data);

      // Check If-None-Match
      const ifNoneMatch = c.req.header("If-None-Match");
      if (ifNoneMatch === etag) {
        return new Response(null, { status: 304 });
      }

      return c.json(JSON.parse(redacted), 200, {
        ETag: etag,
        "Cache-Control": "public, max-age=3600",
      });
    } catch (err) {
      return c.text(`failed to parse jwks: ${(err as Error).message}`, 500);
    }
  });

  return app;
}

File: packages/backlog/internal/relay/certs.go
package relay

import (
	"crypto/ed25519"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"net/http"
	"strings"

	"github.com/yacchi/backlog-cli/packages/backlog/internal/config"
)

func (s *Server) handleRelayCerts(w http.ResponseWriter, r *http.Request) {
	domain := strings.TrimSpace(r.PathValue("domain"))
	if domain == "" {
		http.Error(w, "domain is required", http.StatusBadRequest)
		return
	}

	tenant, ok := findTenantByAllowedDomain(s.cfg.Server().Tenants, domain)
	if !ok {
		http.Error(w, "tenant not found", http.StatusNotFound)
		return
	}

	jwksJSON := strings.TrimSpace(tenant.JWKS)
	if jwksJSON == "" {
		http.Error(w, "tenant jwks is empty", http.StatusInternalServerError)
		return
	}

	redacted, err := redactJWKSPrivateKeys([]byte(jwksJSON))
	if err != nil {
		http.Error(w, fmt.Sprintf("failed to parse jwks: %v", err), http.StatusInternalServerError)
		return
	}

	// ETag生成（コンテンツのSHA256ハッシュ）
	hash := sha256.Sum256(redacted)
	etag := `"` + hex.EncodeToString(hash[:16]) + `"`

	// If-None-Matchヘッダーでキャッシュ検証
	if match := r.Header.Get("If-None-Match"); match == etag {
		w.WriteHeader(http.StatusNotModified)
		return
	}

	// キャッシュヘッダー設定
	w.Header().Set("Content-Type", "application/json")
	SetCacheHeaders(w, CacheTypeLong, s.cfg)
	w.Header().Set("ETag", etag)
	_, _ = w.Write(redacted)
}

func redactJWKSPrivateKeys(input []byte) ([]byte, error) {
	var jwks relayJWKS
	if err := json.Unmarshal(input, &jwks); err != nil {
		return nil, err
	}
	if len(jwks.Keys) == 0 {
		return nil, fmt.Errorf("invalid jwks: keys is missing or empty")
	}

	for i := range jwks.Keys {
		key := &jwks.Keys[i]
		if key.D != "" {
			derivedX, err := derivePublicKeyFromSeed(*key)
			if err != nil {
				return nil, err
			}
			key.X = derivedX
		}
		key.D = ""
	}

	return json.Marshal(jwks)
}

func derivePublicKeyFromSeed(jwk relayJWK) (string, error) {
	if jwk.Kty != "" && jwk.Kty != "OKP" {
		return "", fmt.Errorf("unsupported JWK: kty=%s", jwk.Kty)
	}
	if jwk.Crv != "" && jwk.Crv != "Ed25519" {
		return "", fmt.Errorf("unsupported JWK: crv=%s", jwk.Crv)
	}
	if jwk.D == "" {
		return "", fmt.Errorf("missing private key material for %s", jwk.Kid)
	}
	seed, err := base64.RawURLEncoding.DecodeString(jwk.D)
	if err != nil {
		return "", fmt.Errorf("invalid JWK d for %s: %w", jwk.Kid, err)
	}
	if len(seed) != ed25519.SeedSize {
		return "", fmt.Errorf("invalid JWK seed size for %s: %d", jwk.Kid, len(seed))
	}
	privKey := ed25519.NewKeyFromSeed(seed)
	pubKey := privKey.Public().(ed25519.PublicKey)
	return base64.RawURLEncoding.EncodeToString(pubKey), nil
}

func findTenantByAllowedDomain(tenants map[string]config.ResolvedTenant, domain string) (*config.ResolvedTenant, bool) {
	for _, tenant := range tenants {
		if strings.EqualFold(tenant.AllowedDomain, domain) {
			tenantCopy := tenant
			return &tenantCopy, true
		}
	}
	return nil, false
}
